Add cuSOLVER path for torch.linalg.qr

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57317 Add cuSOLVER path for torch.linalg.lstsq
* #57316 Add CUDA support for torch.ormqr
* #57315 Port CPU torch.ormqr to ATen
* #57314 Fix torch.ormqr for non Fortran-contiguous inputs
* #56257 Fix MAGMA qr for empty batched inputs
* **#56256 Add cuSOLVER path for torch.linalg.qr**
* #56255 Remove size arguments for internal orgqr and geqrf calls
* #56254 Add non-allocating helper function for torch.linalg.qr

Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally. See https://github.com/pytorch/pytorch/pull/56256#issuecomment-821069086.

Performance comparison: https://github.com/pytorch/pytorch/pull/56256#issuecomment-821077712.

Differential Revision: [D27960154](https://our.internmc.facebook.com/intern/diff/D27960154)

IvanYashchuk committed:
Add cuSOLVER path for torch.linalg.qr

Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552

[ghstack-poisoned]

IvanYashchuk commented:
Time spent for running `pytest test/test_ops.py -k 'linalg_qr' --durations=5`.
cuSOLVER:
```
====================================================== slowest 5 durations =======================================================
8.03s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_qr_cuda_complex64
2.67s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_qr_cuda_float32
2.65s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_qr_cpu_complex128
1.73s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_complex128
1.37s call     test/test_ops.py::TestOpInfoCUDA::test_duplicate_method_tests_linalg_qr_cuda_float32
================================= 49 passed, 41 skipped, 12294 deselected, 5 warnings in 31.98s ==================================
```

MAGMA:
```
====================================================== slowest 5 durations =======================================================
39.57s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_complex128
11.12s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_qr_cuda_float64
5.31s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_qr_cuda_float32
5.28s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_qr_cuda_complex64
2.75s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_linalg_qr_cpu_complex128
============================ 49 passed, 41 skipped, 12294 deselected, 5 warnings in 81.28s (0:01:21) =============================
```

IvanYashchuk commented:
Here is MAGMA vs cuSOLVER comparison for non-batched square inputs for modes `'complete', 'reduced', 'r'`:
```
|                          | cuSOLVER, 'complete' | MAGMA, 'complete' | cuSOLVER, 'reduced' | MAGMA, 'reduced' | cuSOLVER, 'r' | MAGMA, 'r' |
|--------------------------|----------------------|-------------------|---------------------|------------------|---------------|------------|
| torch.Size([2, 2])       | 0.084                | 8.0               | 0.0774              | 7.6              | 0.0504        | 3.3        |
| torch.Size([8, 8])       | 0.0877               | 7.6               | 0.0872              | 8.1              | 0.0474        | 3.2        |
| torch.Size([16, 16])     | 0.158                | 7.6               | 0.1569              | 8.3              | 0.1577        | 3.3        |
| torch.Size([32, 32])     | 0.4164               | 7.6               | 0.413               | 8.5              | 0.2835        | 3.3        |
| torch.Size([64, 64])     | 0.9334               | 8.0               | 0.9257              | 8.4              | 0.6559        | 3.3        |
| torch.Size([128, 128])   | 2.0622               | 9.3               | 2.045               | 9.8              | 1.554         | 3.9        |
| torch.Size([256, 256])   | 3.5756               | 12.4              | 3.548               | 12.9             | 2.342         | 5.1        |
| torch.Size([512, 512])   | 8.6611               | 17.4              | 8.593               | 18.7             | 5.797         | 8.3        |
| torch.Size([1024, 1024]) | 23.4609              | 36.9              | 23.342              | 37.4             | 15.196        | 15.6       |
| torch.Size([2048, 2048]) | 92.3197              | 118.7             | 92.247              | 120.1            | 54.483        | 43.9       |
| torch.Size([4096, 4096]) | 497.0645             | 694.1             | 494.418             | 695.7            | 277.952       | 243.5      |
| torch.Size([8192, 8192]) | 3267.1995            | 4603.7            | 3250.727            | 4617.3           | 1713.537      | 1536.7     |
Times are in milliseconds (ms).
```

MAGMA is only faster than cuSOLVER for large size inputs and `mode='r'`. For all other cases cuSOLVER is better.

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.
Ref. https://github.com/pytorch/pytorch/issues/51552, https://github.com/pytorch/pytorch/issues/47953

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.


[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally.


[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1777,7 +1777,7 @@ void linalg_qr_out_helper(const Tensor& input, const Tensor& Q, const Tensor& R,
 }
 
-std::tuple<Tensor, Tensor> _linalg_qr_helper_cpu(const Tensor& input, std::string mode) {
+std::tuple<Tensor, Tensor> _linalg_qr_helper_default(const Tensor& input, std::string mode) {
```
Why "default" and not "cpu"?

IvanYashchuk replied:
We have now `linalg_qr_helper_magma` that uses MAGMA for the QR decomposition, it can't be implemented using `geqrf_stub` + `orgqr_stub`, because `orgqr_stub` only supports cuSOLVER for CUDA inputs. In addition, MAGMA doesn't follow LAPACK API for `geqrf` and `orgqr` operations that together form the QR decomposition. That's why we need to have a separate function for MAGMA.

And we have `_linalg_qr_helper_default` with "_default" and not "_cpu" because this function supports both CPU and CUDA inputs, for CUDA inputs cuSOLVER&cuBLAS is used.

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally. See https://github.com/pytorch/pytorch/pull/56256#issuecomment-821069086.

Performance comparison: https://github.com/pytorch/pytorch/pull/56256#issuecomment-821077712.

Differential Revision: [D27960154](https://our.internmc.facebook.com/intern/diff/D27960154)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally. See https://github.com/pytorch/pytorch/pull/56256#issuecomment-821069086.

Performance comparison: https://github.com/pytorch/pytorch/pull/56256#issuecomment-821077712.

Differential Revision: [D27960154](https://our.internmc.facebook.com/intern/diff/D27960154)

[ghstack-poisoned]

mruberry commented:
Time to start landing the second part of this stack!

@xwang233 would you take a look at this PR in the stack?

xwang233 reviewed (approved):
The PR is very concise and LGTM.

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally. See https://github.com/pytorch/pytorch/pull/56256#issuecomment-821069086.

Performance comparison: https://github.com/pytorch/pytorch/pull/56256#issuecomment-821077712.

Differential Revision: [D27960154](https://our.internmc.facebook.com/intern/diff/D27960154)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.qr"


Using cuSOLVER path with `pytest test/test_ops.py -k 'linalg_qr'
--durations=5` cuts the runtime for these tests by 1 minute locally. See https://github.com/pytorch/pytorch/pull/56256#issuecomment-821069086.

Performance comparison: https://github.com/pytorch/pytorch/pull/56256#issuecomment-821077712.

Differential Revision: [D27960154](https://our.internmc.facebook.com/intern/diff/D27960154)

[ghstack-poisoned]

mruberry reviewed (approved):
Stamped
